Resource methods blog (#677)
pulumi/pulumi-hugo: themes/default/content/blog/resource-methods-for-pulumi-packages/index.md

---
title: "Introducing Resource Methods for Pulumi Packages"
date: 2021-10-11T08:00:00-07:00

> draft: true

Reviewer: If this post needs to go live when the PR is merged, we need to change the draft status.

```suggestion
draft: false
```

meta_desc: It's now possible to provide resource methods from Pulumi Packages
meta_image: pulumi-mlc-packages.png
authors:
    - justin-vanpatten
tags:
    - packages
---

> It's now possible to provide resource methods from Pulumi Packages. Resource methods are similar to functions, but instead of being exposed as top-level functions in a module, methods are exposed as methods on a resource class. This allows for a more object oriented approach to exposing functionality -- operations performed by a resource (that potentially use the resource's state) can now be exposed as methods on the resource. Resource methods can be implemented once, in your language of choice, and made available to users in all Pulumi languages.

Reviewer: ```suggestion
It's now possible to provide resource methods from Pulumi Packages. Resource methods are similar to functions, but instead of being exposed as top-level functions in a module, methods are exposed as methods on a resource class. This allows for a more object-oriented approach to exposing functionality&mdash;operations performed by a resource (that potentially use the resource's state) can now be exposed as methods on the resource. Resource methods can be implemented once, in your language of choice, and made available to users in all Pulumi languages.
```

<!--more-->

> When authoring component resources, it's often useful to provide additional functionality through methods on the component. For example, the `Cluster` component in the [`eks`]({{<relref "/docs/reference/pkg/eks">}}) package has a [`getKubeconfig`](https://github.com/pulumi/pulumi-eks/blob/700d73e961976e58762cb9c723ad2fa838052f46/nodejs/eks/cluster.ts#L1482) method that can be used to generate a kubeconfig for authentication with the cluster that does not use the default AWS credential provider chain, but is instead scoped based on the passed-in arguments. Until now, that method has only been available from JavaScript/TypeScript (the language the `Cluster` component was written in). With the new support for resource methods for Pulumi Packages, we can make this method available to all the Pulumi languages, which is exactly what we've done in [#628](https://github.com/pulumi/pulumi-eks/pull/628).

Author: I'm on PTO this week, but if we get https://github.com/pulumi/pulumi-eks/pull/628 reviewed/merged and released (v0.34.0?), we could point folks to the released version.

Reviewer: Relrefs need spaces in there to get generated correctly. 

```suggestion
When authoring component resources, it's often useful to provide additional functionality through methods on the component. For example, the `Cluster` component in the [`eks`]({{< relref "/docs/reference/pkg/eks" >}}) package has a [`getKubeconfig`](https://github.com/pulumi/pulumi-eks/blob/700d73e961976e58762cb9c723ad2fa838052f46/nodejs/eks/cluster.ts#L1482) method that can be used to generate a kubeconfig for authentication with the cluster that does not use the default AWS credential provider chain, but is instead scoped based on the passed-in arguments. Until now, that method has only been available from JavaScript/TypeScript (the language the `Cluster` component was written in). With the new support for resource methods for Pulumi Packages, we can make this method available to all the Pulumi languages, which is exactly what we've done in [#628](https://github.com/pulumi/pulumi-eks/pull/628).
```

{{< chooser language "typescript,python,csharp,go" >}}

{{% choosable language typescript %}}

```typescript
const cluster = new eks.Cluster("mycluster");
const kubeconfig = cluster.getKubeconfig(...);
```

{{% /choosable %}}
{{% choosable language python %}}

```python
cluster = eks.Cluster("mycluster")
kubeconfig = cluster.get_kubeconfig(...)
```

{{% /choosable %}}
{{% choosable language csharp %}}

```csharp
var cluster = new Cluster("mycluster");
var kubeconfig = cluster.GetKubeconfig(...);
```

{{% /choosable %}}

{{% choosable language go %}}

```go
cluster, err := eks.NewCluster(ctx, "mycluster", nil)
if err != nil {
    return err
}
kubeconfig, err := cluster.GetKubeconfig(ctx, &eks.ClusterGetKubeconfigArgs{...})
if err != nil {
    return err
}
```

{{% /choosable %}}

{{% /chooser %}}

> This post will show how you can provide resource methods from your [component packages]({{<relref "/docs/guides/pulumi-packages">}}).

Reviewer: ```suggestion
This post will show how you can provide resource methods from your [component packages]({{< relref "/docs/guides/pulumi-packages" >}}).
```

## Authoring Methods

> It's always been possible to provide module-level functions from Pulumi Packages. For example, the [`aws`]({{<relref "/docs/reference/pkg/aws">}}) package provides many module-level functions, such as the [`getAmi`]({{<relref "docs/reference/pkg/aws/ec2/getami">}}) function which can be used to get the ID of an existing Amazon Machine Image (AMI). Functions are declared in the package's [schema]({{<relref "/docs/guides/pulumi-packages/schema">}}) and their functionality is implemented in the provider through the `Invoke` RPC.

Reviewer: ```suggestion
It's always been possible to provide module-level functions from Pulumi Packages. For example, the [`aws`]({{< relref "/docs/reference/pkg/aws" >}}) package provides many module-level functions, such as the [`getAmi`]({{< relref "/docs/reference/pkg/aws/ec2/getami" >}}) function, which can be used to get the ID of an existing Amazon Machine Image (AMI). Functions are declared in the package's [schema]({{< relref "/docs/guides/pulumi-packages/schema" >}}), and their functionality is implemented in the provider through the `Invoke` remote procedure call (RPC).
```

Methods are authored in a similar manner to functions. Methods are declared in the schema and implemented in the provider's `Call` RPC (similar to `Invoke`).

> Let's walk through an example. We'll author a `Message` component which accepts a message as input and provides a `getMessage` method that accepts a recipient name and returns the message customized for the recipient. To get started authoring a component package, refer to the [package]({{<relref "/docs/guides/pulumi-packages">}}) documentation.

Reviewer: Patching the relref and breaking up the sentence. Some readers will have trouble comprehending the longer sentence, so this reword makes it easier for everyone to follow the conversation.

```suggestion
Let's walk through an example. We'll author a `Message` component that accepts a message as input. The component then provides a `getMessage` method that accepts a recipient name and returns the message customized for the recipient. To get started authoring a component package, refer to the [package]({{< relref "/docs/guides/pulumi-packages" >}}) documentation.
```

### Schema

> We'll start with declaring the method and component in the [Pulumi schema]({{<relref "/docs/guides/pulumi-packages/schema">}}). First, define the function representing the method:

Reviewer: ```suggestion
We'll start with declaring the method and component in the [Pulumi schema]({{< relref "/docs/guides/pulumi-packages/schema" >}}). First, define the function representing the method:
```

          }
        },
        "required": ["__self__", "recipient"]
      },
      "outputs": {
        "properties": {
          "result": {
            "type": "string"
          }
        },
        "required": ["result"]
      }
    }
  },
```

Our method has two required arguments: `__self__` and `recipient`. `__self__` is a special input required by all methods that represents the component resource and is typed as such. The method has one result named `result`.

Next, define our `Message` component:

```json
  "resources": {
    "example:index:Message": {
      "isComponent": true,
      "inputProperties": {
        "message": {
          "type": "string"
        },
      },
      "requiredInputs": ["message"],
      "properties": {
        "message": {
          "type": "string"
        },
      },
      "required": ["message"],
      "methods": {
        "getMessage": "example:index:Message/getMessage"
      }
    }
  },
```

Our component has a single required input/output property: `message`. The method is specified in the `methods` property, which references the method's function definition: `"getMessage": "example:index:Message/getMessage"`.

> Altogether our schema looks like:

Reviewer: Updating language to improve accessibility (some folks will be on screen readers).

```suggestion
Here's our schema all together:
```

          },
          "recipient": {
            "type": "string"
          }
        },
        "required": ["__self__", "recipient"]
      },
      "outputs": {
        "properties": {
          "result": {
            "type": "string"
          }
        },
        "required": ["result"]
      }
    }
  },
  "resources": {
    "example:index:Message": {
      "isComponent": true,
      "inputProperties": {
        "message": {
          "type": "string"
        },
      },
      "requiredInputs": ["message"],
      "properties": {
        "message": {
          "type": "string"
        },
      },
      "required": ["message"],
      "methods": {
        "getMessage": "example:index:Message/getMessage"
      }
    }
  },
  "language": {
    "csharp": {
      "packageReferences": {
        "Pulumi": "3.12"
      },
      "liftSingleValueMethodReturns": true
    },
    "go": {
      "liftSingleValueMethodReturns": true
    },
    "nodejs": {
      "devDependencies": {
        "@types/node": "latest"
      },
      "liftSingleValueMethodReturns": true
    },
    "python": {
      "liftSingleValueMethodReturns": true
    }
  }
}
```

> To make it easier for users of our method, we also set `"liftSingleValueMethodReturns": true` for each language, which makes single value methods return the single value directly, rather than wrapping the results in a `Result` type.

Reviewer: ```suggestion
To make it easier for users of our method, we also set `"liftSingleValueMethodReturns": true` for each language, which makes single-value methods return the single value directly, rather than wrapping the results in a `Result` type.
```

### Component Implementation

> We can implement the component in TypeScript/JavaScript, Python, or Go, and make it available to any Pulumi language.

Reviewer: I know we don't have a C# example right now. Feasibly, someone could use C#, right? We probably should adjust the wording here as follows to avoid someone asking why the system isn't available in C#:

```suggestion
We can implement the component and make it available to any Pulumi language. We're going to show some implementation examples in TypeScript/JavaScript, Python, and Go.
```

Here's the implementation of the `Message` component:

> {{< chooser language "typescript,python,go" >}}

Reviewer: Use note: Since there's C# in the chooser further up this blog post, the C# user here will default to TypeScript. It might be better to drop C# above in the chooser if we don't have an example for it here to avoid confusion.

        return pulumi.Output.concat(recipient, ", ", self.message, "!")
```

{{% /choosable %}}
{{% choosable language go %}}

```go
import (
	"errors"

	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

type Message struct {
	pulumi.ResourceState
	Message  pulumi.StringOutput `pulumi:"message"`
}

type MessageArgs struct {
	Message  pulumi.StringInput `pulumi:"message"`
}

func NewMessage(ctx *pulumi.Context, name string, args *MessageArgs, opts ...pulumi.ResourceOption) (*Message, error) {
	if args == nil {
		return nil, errors.New("args is required")
	}

	message := &Message{}
	err := ctx.RegisterComponentResource("example:index:Message", name, message, opts...)
	if err != nil {
		return nil, err
	}

	message.Message = args.Message.ToStringOutput()
	if err := ctx.RegisterResourceOutputs(message, pulumi.Map{
		"message":  args.Message,
	}); err != nil {
		return nil, err
	}

	return message, nil
}

type GetMessageArgs struct {
	Recipient pulumi.StringInput `pulumi:"recipient"`
}

func (c *Message) GetMessage(args *GetMessageArgs) StringOutput {
	return pulumi.Sprintf("%s, %s!", args.Recipient, c.Message)
}
```

{{% /choosable %}}

{{% /chooser %}}

### Provider RPCs

> Next, wire up the provider RPCs with the component implementation.

Reviewer: ```suggestion
Next, wire up the provider RPCs with the component implementation:
```

```typescript
const component = new example.Message("mycomponent", {
    message: "hello world",
});
export const message = component.getMessage({ recipient: "Alice" }); // Exports "Alice, hello world!"
```

{{% /choosable %}}
{{% choosable language python %}}

```python
component = example.Message("mycomponent", message="hello world")
message = component.get_message(recipient="Alice")
pulumi.export("message", message) # Exports "Alice, hello world!"
```

{{% /choosable %}}
{{% choosable language csharp %}}

```csharp
var component = new Message("mycomponent", new MessageArgs
{
    Message = "hello world",
});
// Exports "Alice, hello world!"
this.Message = component.GetMessage(new MessageGetMessageArgs
{
    Recipient = "Alice",
});
```

{{% /choosable %}}

{{% choosable language go %}}

```go
component, err := example.NewMessage(ctx, "mycomponent", &example.MessageArgs{
	Message: "hello world",
})
if err != nil {
	return err
}
message, err := component.GetMessage(ctx, &example.MessageGetMessageArgs{Recipient: "Alice"})
if err != nil {
	return err
}
pulumi.Export("message", message) // Exports "Alice, hello world!"
```

{{% /choosable %}}

{{% /chooser %}}

## Wrapping Up

> With support for resource methods for Pulumi Packages, you can now create component resources with methods, and make them available to use from any Pulumi language. We look forward to seeing the components you create!

Reviewer: ```suggestion
With support for resource methods for Pulumi Packages, you can now create component resources with methods and make them available to use from any Pulumi language. We look forward to seeing the components you create!
```
